Radis download Error for N2

### 🐛 Describe the bug

from radis import calc_spectrum
s = calc_spectrum(wavelength_min=1000,
                  wavelength_max=20000,
                  Tgas=300,
                  pressure=1,
                  molecule='N2',
                  optimization=None,
                  cutoff=1e-23,
                  isotope='2')
s.plot('abscoeff', wunit='nm')

I tried calculating the absorption coefficient of N2 through the code above. However, it appears that a problem occurred during the download process.

AssertionError                            Traceback (most recent call last)
~\AppData\Local\Temp\ipykernel_12248\3939192323.py in <module>
      1 from radis import calc_spectrum
----> 2 s = calc_spectrum(wavelength_min=1000,
      3                   wavelength_max=20000,
      4                   Tgas=300,
      5                   pressure=1,

c:\ANACONDA\lib\site-packages\radis\lbl\calc.py in calc_spectrum(wmin, wmax, wunit, Tgas, Tvib, Trot, pressure, molecule, isotope, mole_fraction, diluent, path_length, databank, medium, wstep, truncation, neighbour_lines, cutoff, parsum_mode, optimization, chunksize, broadening_method, overpopulation, name, save_to, use_cached, mode, export_lines, verbose, return_factory, **kwargs)
    532         kwargs_molecule.update(**dict_arguments)
    533 
--> 534         generated_spectrum = _calc_spectrum_one_molecule(
    535             wavenum_min=wavenum_min,
    536             wavenum_max=wavenum_max,

c:\ANACONDA\lib\site-packages\radis\lbl\calc.py in _calc_spectrum_one_molecule(wavenum_min, wavenum_max, input_wunit, Tgas, Tvib, Trot, pressure, overpopulation, molecule, isotope, mole_fraction, path_length, databank, medium, wstep, truncation, neighbour_lines, cutoff, parsum_mode, optimization, chunksize, broadening_method, name, use_cached, verbose, mode, export_lines, return_factory, diluent, **kwargs)
    798 
    799         # Finally, LOAD :
--> 800         sf.fetch_databank(**conditions)
    801     elif exists(databank):
    802         conditions = {

c:\ANACONDA\lib\site-packages\radis\lbl\loader.py in fetch_databank(self, source, database, parfunc, parfuncfmt, levels, levelsfmt, load_energies, include_neighbouring_lines, parse_local_global_quanta, drop_non_numeric, db_use_cached, lvl_use_cached, memory_mapping_engine, load_columns, parallel, extra_params)
   1161                     isotope_list = ",".join([str(k) for k in self._get_isotope_list()])
...
    252         df["branch"] = new_col
    253 
--> 254     assert df.dtypes["branch"] == np.int64

AssertionError: 

This error message keeps occurring. I am wondering if there is a problem downloading data from Hitran or if there is a problem with the code. Can you please tell me how to troubleshoot this? Thank you for answer.

### 💡 Possible solutions

_No response_

### 🎲 Radis version

0.15

### 💻 Operating system

Windows

AssertionError                            Traceback (most recent call last)
Cell In[13], line 1
----> 1 s = calc_spectrum(wavelength_min=1000,
      2                   wavelength_max=20000,
      3                   Tgas=300,
      4                   pressure=1,
      5                   molecule='N2',
      6                   optimization=None,
      7                   cutoff=1e-23,
      8                   isotope='1')
      9 s.plot('abscoeff', wunit='nm')

File ~\AppData\Local\anaconda3\envs\radis_env\lib\site-packages\radis\lbl\calc.py:534, in calc_spectrum(wmin, wmax, wunit, Tgas, Tvib, Trot, pressure, molecule, isotope, mole_fraction, diluent, path_length, databank, medium, wstep, truncation, neighbour_lines, cutoff, parsum_mode, optimization, chunksize, broadening_method, overpopulation, name, save_to, use_cached, mode, export_lines, verbose, return_factory, **kwargs)
    531 # We add all of the DICT_INPUT_ARGUMENTS values:
    532 kwargs_molecule.update(**dict_arguments)
--> 534 generated_spectrum = _calc_spectrum_one_molecule(
    535     wavenum_min=wavenum_min,
    536     wavenum_max=wavenum_max,
    537     input_wunit=input_wunit,
    538     Tgas=Tgas,
    539     Tvib=Tvib,
    540     Trot=Trot,
    541     pressure=pressure,
    542     # overpopulation=overpopulation,  # now in dict_arguments
    543     molecule=molecule,
    544     # isotope=isotope,                # now in dict_arguments
    545     # mole_fraction=mole_fraction,    # now in dict_arguments
    546     path_length=path_length,
    547     # databank=databank,              # now in dict_arguments
    548     medium=medium,
    549     wstep=wstep,
    550     truncation=truncation,
    551     neighbour_lines=neighbour_lines,
    552     cutoff=cutoff,
    553     parsum_mode=parsum_mode,
    554     optimization=optimization,
    555     chunksize=chunksize,
    556     broadening_method=broadening_method,
    557     name=name,
    558     use_cached=use_cached,
    559     verbose=verbose,
    560     mode=mode,
    561     export_lines=export_lines,
    562     return_factory=return_factory,
    563     diluent=diluent,
    564     **kwargs_molecule,
    565 )
    566 if return_factory:
    567     factory_dict[molecule] = generated_spectrum[1]

File ~\AppData\Local\anaconda3\envs\radis_env\lib\site-packages\radis\lbl\calc.py:800, in _calc_spectrum_one_molecule(wavenum_min, wavenum_max, input_wunit, Tgas, Tvib, Trot, pressure, overpopulation, molecule, isotope, mole_fraction, path_length, databank, medium, wstep, truncation, neighbour_lines, cutoff, parsum_mode, optimization, chunksize, broadening_method, name, use_cached, verbose, mode, export_lines, return_factory, diluent, **kwargs)
    797     conditions["parse_local_global_quanta"] = (not _equilibrium) or export_lines
    799     # Finally, LOAD :
--> 800     sf.fetch_databank(**conditions)
    801 elif exists(databank):
    802     conditions = {
    803         "path": databank,
    804         "drop_columns": drop_columns,
   (...)
    807         "db_use_cached": use_cached,
    808     }

File ~\AppData\Local\anaconda3\envs\radis_env\lib\site-packages\radis\lbl\loader.py:1163, in DatabankLoader.fetch_databank(self, source, database, parfunc, parfuncfmt, levels, levelsfmt, load_energies, include_neighbouring_lines, parse_local_global_quanta, drop_non_numeric, db_use_cached, lvl_use_cached, memory_mapping_engine, load_columns, parallel, extra_params)
   1160 else:
   1161     isotope_list = ",".join([str(k) for k in self._get_isotope_list()])
-> 1163 df, local_paths = fetch_hitran(
   1164     molecule,
   1165     isotope=isotope_list,
   1166     local_databases=join(local_databases, "hitran"),
   1167     load_wavenum_min=wavenum_min,
   1168     load_wavenum_max=wavenum_max,
   1169     columns=columns,
   1170     cache=db_use_cached,
   1171     verbose=self.verbose,
   1172     return_local_path=True,
   1173     engine=memory_mapping_engine,
   1174     parallel=parallel,
   1175     extra_params=extra_params,
   1176 )
   1177 self.params.dbpath = ",".join(local_paths)
   1179 # ... explicitely write all isotopes based on isotopes found in the database

File ~\AppData\Local\anaconda3\envs\radis_env\lib\site-packages\radis\io\hitran.py:184, in fetch_hitran(molecule, extra_params, local_databases, databank_name, isotope, load_wavenum_min, load_wavenum_max, columns, cache, verbose, clean_cache_files, return_local_path, engine, output, parallel, parse_quanta)
    182 download_files = ldb.get_missing_files(local_file)
    183 if download_files:
--> 184     ldb.download_and_parse(download_files, cache=cache, parse_quanta=parse_quanta)
    186 # Register
    187 if not ldb.is_registered():

File ~\AppData\Local\anaconda3\envs\radis_env\lib\site-packages\radis\api\hitranapi.py:1319, in HITRANDatabaseManager.download_and_parse(self, local_file, cache, parse_quanta)
   1297 df = pd.DataFrame(LOCAL_TABLE_CACHE[data_file.split(".")[0]]["data"])
   1298 df.rename(
   1299     columns={
   1300         "molec_id": "id",
   (...)
   1317     inplace=True,
   1318 )
-> 1319 df = post_process_hitran_data(
   1320     df,
   1321     molecule=molecule,
   1322     parse_quanta=parse_quanta,
   1323 )
   1325 wmin_final = min(wmin_final, df.wav.min())
   1326 wmax_final = max(wmax_final, df.wav.max())

File ~\AppData\Local\anaconda3\envs\radis_env\lib\site-packages\radis\api\hitranapi.py:398, in post_process_hitran_data(df, molecule, verbose, drop_non_numeric, parse_quanta)
    396 if drop_non_numeric:
    397     if "branch" in df:
--> 398         replace_PQR_with_m101(df)
    399     df = drop_object_format_columns(df, verbose=verbose)
    401 return df

File ~\AppData\Local\anaconda3\envs\radis_env\lib\site-packages\radis\api\tools.py:254, in replace_PQR_with_m101(df)
    251     new_col = df["branch"].replace(["P", "Q", "R"], [-1, 0, 1])
    252     df["branch"] = new_col
--> 254 assert df.dtypes["branch"] == np.int64

AssertionError: 
Yeah the same happens for me
Did you get the solution

--- a/radis/api/tools.py
+++ b/radis/api/tools.py
@@ -254,7 +254,7 @@ def drop_object_format_columns(df, verbose=True):
 
 
 def replace_PQR_with_m101(df):
-    """Return P, Q, R in column ``branch`` with -1, 0, 1 to get a fully numeric
+    """Return O, P, Q, R, S in column ``branch`` with -2, -1, 0, 1, 2 to get a fully numeric
     database. This improves performances quite a lot, as Pandas doesnt have a
     fixed-string dtype hence would use the slow ``object`` dtype.
 
@@ -288,7 +288,7 @@ def replace_PQR_with_m101(df):
 
     # Then do the actual mapping
     if df.dtypes["branch"] != np.int64:
-        mapping = {"P": -1, "Q": 0, "R": 1}
+        mapping = {"P": -1, "Q": 0, "R": 1, "O": -2, "S": 2}
         if df_type == pd.DataFrame:  # pandas
             new_col = df["branch"].replace(mapping)
         else:  # vaex

